File: test/believe/util/UtilTest.java
package believe.util;

import static believe.util.MapEntry.entry;
import static org.hamcrest.Matchers.containsInAnyOrder;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.nullValue;
import static org.junit.Assert.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;
import static org.mockito.MockitoAnnotations.initMocks;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.newdawn.slick.Graphics;
import org.newdawn.slick.geom.Rectangle;

public class UtilTest {
  @Mock private Graphics g;
  @Mock private believe.geometry.Rectangle childRect;
  @Mock private Rectangle parentRect;

  @Before
  public void setUp() {
    initMocks(this);
  }

  @Test
  public void changeClipContextShouldReturnIntersectionOfParentAndChildClips() {
    final believe.geometry.Rectangle newClip = new believe.geometry.Rectangle(50, 50, 50, 50);
    final float x = 0.0f;
    final float y = 0.0f;
    final float width = 100.0f;
    final float height = 100.0f;

    when(g.getClip()).thenReturn(parentRect);
    when(parentRect.getX()).thenReturn(x);
    when(parentRect.getY()).thenReturn(y);
    when(parentRect.getWidth()).thenReturn(width);
    when(parentRect.getHeight()).thenReturn(height);
    when(childRect.intersection(any())).thenReturn(newClip);

    Rectangle oldClip = Util.changeClipContext(g, childRect);
    assertThat(oldClip.getX(), is(x));
    assertThat(oldClip.getY(), is(y));
    assertThat(oldClip.getWidth(), is(width));
    assertThat(oldClip.getHeight(), is(height));
  }

  @Test
  public void changeClipContextShouldReturnChildClipIfParentClipIsNull() {
    when(g.getClip()).thenReturn(null);
    Rectangle oldClip = Util.changeClipContext(g, childRect);
    assertThat(oldClip, nullValue());
  }

  @Test
  public void resetClipContextShouldSetGraphicsClipToGivenRectangle() {
    Util.resetClipContext(g, parentRect);
  }

  @Test
  public void hashSetOfReturnsProperHashSet() {
    assertThat(
        Util.hashSetOf("one", "two", "three"), containsInAnyOrder("one", "two", "three"));
  }

  @Test(expected = UnsupportedOperationException.class)
  public void immutableSetOfCannotBeModified() {
    Set<String> set = Util.immutableSetOf("value");
    set.add("some other value");
  }

  @Test
  public void hashMapOfReturnsCorrectHashMap() {
    HashMap<String, String> expectedMap = new HashMap<>();
    expectedMap.put("key", "value");
    expectedMap.put("key2", "value2");
    assertThat(Util.hashMapOf(entry("key", "value"), entry("key2", "value2")), is(expectedMap));

    HashMap<String, String> emptyMap = new HashMap<>();
    assertThat(Util.hashMapOf(), is(emptyMap));
  }

  @Test(expected = UnsupportedOperationException.class)
  public void immutableMapOfCannotBeModified() {
    Map<String, String> map = Util.immutableMapOf(entry("key", "value"));
    map.put("key", "value2");
  }
}

File: javatests/believe/util/UtilTest.kt
package believe.util

import believe.util.MapEntry.Companion.entry
import believe.util.Util.changeClipContext
import believe.util.Util.hashMapOf
import believe.util.Util.hashSetOf
import believe.util.Util.immutableMapOf
import believe.util.Util.immutableSetOf
import com.google.common.truth.Truth.assertThat
import com.nhaarman.mockitokotlin2.check
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.junit.jupiter.api.Test
import org.mockito.Mockito.verify
import org.newdawn.slick.Graphics
import org.newdawn.slick.geom.Rectangle

class UtilTest {
    private val g: Graphics = mock()

    @Test
    fun changeClipContext_returnsOldClipAndSetsNewClipOnGraphics() {
        val expectedParentClip = Rectangle(0f, 0f, 100f, 100f)
        val newClip = believe.geometry.Rectangle(50f, 50f, 100f, 100f)
        whenever(g.clip).thenReturn(expectedParentClip)

        val actualParentClip = g.changeClipContext(newClip)

        with(actualParentClip!!) {
            assertThat(x).isEqualTo(0f)
            assertThat(y).isEqualTo(0f)
            assertThat(width).isEqualTo(100f)
            assertThat(height).isEqualTo(100f)
        }
        verify(g).clip = check {
            with(it) {
                assertThat(x).isEqualTo(50f)
                assertThat(y).isEqualTo(50f)
                assertThat(width).isEqualTo(100f)
                assertThat(height).isEqualTo(100f)
            }
        }
    }

    @Test
    fun changeClipContext_parentIsNull_returnsNull() {
        whenever(g.clip).thenReturn(null)

        val oldClip = g.changeClipContext(believe.geometry.Rectangle(123f, 123f, 123f, 123f))

        assertThat(oldClip).isNull()
    }

    @Test
    fun hashSetOf_returnsProperHashSet() {
        assertThat(hashSetOf("one", "two", "three")).containsAllOf("one", "two", "three")
    }

    @Test
    fun immutableSetOf_cannotBeModified() {
        assertThat(immutableSetOf("one", "two", "three")).containsAllOf("one", "two", "three")
    }

    @Test
    fun hashMapOf_returnsCorrectHashMap() {
        val expectedMap = hashMapOf(
            Pair("key", "value"), Pair("key2", "value2")
        )

        assertThat(
            hashMapOf(
                entry("key", "value"), entry("key2", "value2")
            )
        ).containsExactlyEntriesIn(expectedMap)
    }

    @Test
    fun immutableMapOf_cannotBeModified() {
        val expectedMap = hashMapOf(
            Pair("key", "value"), Pair("key2", "value2")
        )

        assertThat(
            immutableMapOf(
                entry("key", "value"), entry("key2", "value2")
            )
        ).containsExactlyEntriesIn(expectedMap)
    }
}